README.md
# virtual-invention-dice
Inspired from Atomic Shrimp's invention cube. Also serves as a simpler exercise for integrating PWA. For more information about the inspiration, you can read more about it in [this link](http://atomicshrimp.com/post/2014/01/20/Invention-Dice). There's also a video from the creator of this idea which you can see it [right here](https://www.youtube.com/watch?v=NBdVpiWUKhU).

## Demo
![Demo of the app](./docs/assets/demo.gif)

## Features of the app
- Defining your own set of virtual idea dice set (or VIDS, as I call them)
- Storing VIDS in IndexedDB (with [localforage](https://localforage.github.io/localForage/))
- Import and export of VIDS
- Saving dice results in IndexedDB (with [localforage](https://localforage.github.io/localForage/))

## Development build
If you want to open this app locally on your machine, just do the following:

- clone this repo or download the zip of this repo provided
- install all of the dependencies needed (`npm i`)
- open the app over `localhost` with `npm run serve`

Voila! It's open now! 😁

If you want to contribute or just add features that you want, take note of the following configurations:

- created with [Vue CLI](https://cli.vuejs.org/guide/) and the [official Vue router library](https://router.vuejs.org/)
- uses [Tailwind CSS](http://tailwindcss.com/) for most of the CSS
- uses [ESLint](https://github.com/eslint/eslint) and it has the [standard style guide](https://github.com/standard/standard) as the configuration
- all of the usual configurations and constants like the routes, app information, and the author are located in [`./src/appConstants.js`](./src/appConstants.js), I've made it so that I can update it easily in one place

Also, for future references:
- Node v11.3.0
- npm 6.4.1

## Todo
- Multiple virtual idea dice set (VIDS) support


docs/CHANGELOG.adoc
= Changelog
All notable changes to this project will be documented in this file since 
July 11 of 2019 (2019-07-11).

The format is based on https://keepachangelog.com/en/1.0.0/[Keep a Changelog],
and this project adheres to https://semver.org/spec/v2.0.0.html[Semantic Versioning].

== [Unreleased]

== [0.3.1] - 2019-07-16
=== Changed
* Improved the UI/UX of form inputs.

* Classes and components has been changed/refactored. 
Here's a list of notable changes:
** Classes has been restructured to make the output data less 
bloated.
** Die action buttons are grouped. This allows for less clutter 
while being able to introduce more functions in the future.
** Form inputs now works as expected on 'Enter' keypress.
** Textareas are now used instead of the label/input pair. 
** Dynamic grid layout on die items has been 'improved' 
(by measuring the length of the input, lol). This is also needed 
now that `<textarea>` is being used.

== [0.3.0] - 2019-07-15
=== Added
* A help section in the main app page.
** You can also access the help section with `F1`.

* New app functionalities for locking dice sets:
** Locking all of the dice
** Unlocking all of the dice
** Toggling all of the dice lock

* New keyboard shortcuts to the main app page: 
** `Ctrl + Shift + M` for adding a new die item on the currently shown die.
** `Ctrl + Shift + L` for locking all of the dice regardless of their lock status.
** `Ctrl + Shift + :` for unlocking all of the dice regardless of their lock status.
** `Ctrl + Shift + "` for toggling all of the dice locks (unlock if the die is locked, vice versa).

=== Changed
* Revise the UI (for the second time). Here's the detail of some 
of the notable UI changes:
** Change the text into icons in the buttons. It also cause the main 
UI to be shorter (which means it's also a less pain to use it in a mobile 
device).
** Also, the buttons are now into a consistent button group.
** The "Add die item" button in the die component moved inside of the die popover.
Now it feels like an proper group of actions in the popover.
** The die component also got a overhaul with the "fancy" two column into a 
simple and more comprehensive one column on bigger screens similar to the mobile 
screen format.

* The "new die" keyboard shortcut now maps to `Ctrl + Shift + N` 
(yes, that was stupid of me...)

== [0.2.0] - 2019-07-14
=== Added
* Locking function for the die in case you want to not accidentally modify the die 
(i.e. adding/removing items, editing items) or something.

* Disabling dice now prevents you from modifying anything in the die 
(it's not a bug, it's a feature, I swear...). It basically functions as a lock 
function except it doesn't count towards any of the app's operations as if it 
was deleted or never existed in the first place.

* Set a maximum length set for die name, die items, and idea set names. Here's the 
following specifications:
** For die and idea set names, *its length must be less than 64 characters*.
** For die items, *its length must be less than 32 characters*.
** There's no format for the required characters. You can put all of the exclamation marks, 
emojis, and semicolons all you want.

* 'Undo' function for previously deleted dice (you can only retrieve up to 20 recently 
deleted dice at any given time). Take note that you can only hold it in the current session. 
Changing pages or reloading will lose the deleted dice stack.

* New keyboard shortcuts to the main app page
** `Shift + N` for a new die
** `Ctrl + G` for generating an idea set
** `Ctrl + D` for clearing the generated output
** `Ctrl + S` for saving the output in the idea set database
** `Ctrl + I` for importing dice set file
** `Ctrl + E` for exporting the current dice set
** `Ctrl + Delete` for deleting the current die shown in the screen
** `Ctrl + Z` for adding back the most recent deleted die

=== Changed
* Unique ID format has been revised to base 36 to produce less data while being able 
to generate more possible combinations. It should have no effect on previously 
generated data.
** It also makes the newly generated items to have shorter names so it'll have less 
hassle of growing more room in the GUI.

* Minor refactor of components code.

* A minor revamp of the UI and improved the UX (hopefully...).
** A big detail that I want to point out here is how the remove die item button. Now 
it's constantly present with the item since it produces weird quirks 
and it just feels unnatural on the UX side.

* Use of https://tailwindcss.com[Tailwind CSS] 1.0 therefore 
https://tailwindcss.com/docs/release-notes/#tailwind-css-v1-0[big changes happened to its Tailwind config file].
** A minor site aesthetic change has also occured as a side effect of the major upgrade due to the new approach
of Tailwind CSS. Also, because of that, the UX has also improved.


src/constants.js
import localforage from 'localforage'

export const appName = 'Virtual Idea Dice'
export const author = {
  name: 'Gabriel Arazas',
  alias: 'foo-dogsquared',
  link: 'https://foo-dogsquared.github.io/',
  socialLinks: {
    github: 'https://github.com/foo-dogsquared/',
    twitter: 'https://twitter.com/foo_dogsquared',
    linkedin: 'https://linkedin.com/in/gabriel-arazas'
  }
}

export const appConfig = {
  db: {
    name: 'virtual-idea-dice',
    dieStoreName: 'dice',
    ideasStoreName: 'ideas'
  },
  name: 'Virtual Idea Dice',
  icon: '🎲',
  tagline: 'Explore randomly generated ideas with your mind and imagination.',
  repo: 'https://github.com/foo-dogsquared/virtual-idea-dice'
}

export const appHighlights = [
  {
    header: `Inspired from Atomic Shrimp's Invention Dice`,
    description: [
      `This app is inspired from <a href="https://www.youtube.com/watch?v=NBdVpiWUKhU" target="_blank">Atomic Shrimp's Invention Dice</a>.`,
      `In this case, I made the app to be fully customizable for the user in order to explore with their own set of attributes that they've put together.`
    ],
    img: 'app'
  },
  {
    header: 'Create your own set of dice',
    description: [
      "It's like Lego but with dice. Want to create your idea dice set for games? Why not!",
      "What about food? Places? Programs? Hobbies? Or anything? That's your stuff to make!",
      'The only limit here is your imagination (and your categorizing skills).'
    ],
    img: 'create'
  },
  {
    header: 'Share your idea dice set',
    description: [
      `It's possible to share your virtual idea dice set (or VIDS as I like to call them) with others through exporting and importing of your VIDS.`
    ],
    img: 'share'
  },
  {
    header: 'Almost like a native mobile app',
    description: [
      `It's built with mobile-first approach so it can feel like a native mobile app (and skip the developer fees for me) which you can do by adding the app to the home screen.`,
      `Also it works entirely offline once you visited this app (which you have already done 😁). Go on and give it a try.`
    ],
    img: 'phone'
  },
  {
    header: "It's free (and open source) real software",
    description: [
      `Its code is free and open to everyone which you can view it right <a href=${appConfig.repo} target="_blank">here</a>.`,
      `Which means you can help improve this project whether by suggestions, criticisms, or code.`
    ],
    img: 'free'
  }
]

export const routes = {
  home: {
    path: '/',
    name: 'home'
  },
  gettingStarted: {
    path: '/getting-started',
    name: 'getting started'
  },
  app: {
    path: '/app',
    name: 'app'
  },
  ideas: {
    path: '/saved-ideas',
    name: 'saved ideas'
  }
}

export const colors = {
  'brand-color-1': '#17a398',
  'brand-color-dark': '#086375',
  'brand-color-light': '#bfd7ea',
  'black-transparent': 'rgba(0, 0, 0, 0.3)'
}

export const diceStorageKey = 'virtual-idea-dice'

export const state = {
  dieRenaming: 'die-renaming',
  itemEditing: 'item-editing',
  ideaSetRenaming: 'idea-set-renaming'
}

// ID-related functions
export const NUMBER_SYSTEM = 36
export const MAXIMUM_NUMBER_OF_DIGITS = 5
export const MAXIMUM_NUMBER_OF_COMBINATIONS = Math.pow(NUMBER_SYSTEM, MAXIMUM_NUMBER_OF_DIGITS)

export function generateId () {
  const code = Number(Math.floor(Math.random() * MAXIMUM_NUMBER_OF_COMBINATIONS)).toString(NUMBER_SYSTEM)
  return code
}

export function convertToProgramNumberSystem (number) {
  if (typeof number === 'number') {
    return number.toString(NUMBER_SYSTEM)
  } else if (typeof number === 'string' && !Number(`0x${number}`)) {
    return generateId()
  }
}

export const diceStorageInstance = localforage.createInstance({
  storeName: appConfig.db.dieStoreName,
  name: diceStorageKey
})

export const ideasStorageInstance = localforage.createInstance({
  storeName: appConfig.db.ideasStoreName,
  name: diceStorageKey
})

export const MAXIMUM_NAME_LENGTH = 256
export const DIE_NAME_MAXIMUM_LENGTH = 64

export class Die {
  /**
   * A die object that holds die item that represents faces of the die.
   * @param {Object} object - a destructured object that contains the option for
   *                          building the Die
   */
  constructor ({ id = generateId(), name = id, items = [], enabled = true, locked = false } = {}) {
    this.name = String(name)
    // trimming the name if it goes over the maximum length
    this.trimExtraCharacters()

    if (typeof id === 'number') this.id = convertToProgramNumberSystem(id)
    else this.id = String(id)

    this.items = []
    for (const dieItem of items) {
      if (dieItem instanceof DieItem) {
        this.items.push(dieItem)
      } else if (dieItem instanceof Object && dieItem['name'] && dieItem['itemId']) {
        const _dieItem = new DieItem(dieItem)
        this.items.push(_dieItem)
      }
    }

    this.enabled = enabled
    this.locked = locked
  }

  addDieItem (dieItem = undefined) {
    if (!this.enabled || this.locked) return

    let _dieItem

    if (dieItem instanceof DieItem) _dieItem = dieItem
    else _dieItem = new DieItem({ die: this })

    this.items.push(_dieItem)

    return _dieItem
  }

  removeDieItem (item) {
    if (!this.enabled || this.locked) return

    const itemIndex = this.items.indexOf(item)
    this.items.splice(itemIndex, 1)
  }

  checkLockOrEnabled () {
    if (!this.enabled || this.locked) return false

    return true
  }

  toggleDisabling () {
    this.enabled = !this.enabled
  }

  toggleLock () {
    this.locked = !this.locked
  }

  trimExtraCharacters () {
    if (this.name.length > DIE_NAME_MAXIMUM_LENGTH) {
      this.name = this.name.slice(0, DIE_NAME_MAXIMUM_LENGTH)
    }

    this.name = this.name.trim()
  }
}

export const DIE_ITEM_NAME_MAX_LENGTH = 64

export class DieItem {
  /**
   * A die item is a basically like a face on the die object.
   * @param {Object} options
   */
  constructor ({ name = undefined, id = generateId(), die = undefined }) {
    if (name) {
      this.name = name
    } else if (!name && die && die instanceof Die) {
      this.name = `${die.name}-${id}`
    } else {
      this.name = 'New Die Item'
    }

    this.itemId = id
  }

  trimExtraCharacters () {
    if (this.name.length > DIE_ITEM_NAME_MAX_LENGTH) {
      this.name = this.name.slice(0, DIE_ITEM_NAME_MAX_LENGTH)
    }

    this.name = this.name.trim()
  }
}

export const IDEA_SET_NAME_MAX_LENGTH = 64

export class IdeaSet {
  /**
   * An idea set made up of the results from 'rolling of the dice set'.
   * @param {Object} options - an object that specifies options for building
   *
   */
  constructor ({ id = generateId(), name = `Idea Set #${id}`, shards = undefined } = {}) {
    this.id = id
    this.name = name
    this.shards = shards || []
  }

  /**
   * Adds an idea shard to the set.
   * @param {IdeaShard} ideaShard - the idea shard to be added
   */
  addShard (ideaShard) {
    if (!(ideaShard instanceof IdeaShard)) return
    this.shards.push(ideaShard)
  }

  trimExtraCharacters () {
    if (this.name.length > IDEA_SET_NAME_MAX_LENGTH) {
      this.name = this.name.slice(0, IDEA_SET_NAME_MAX_LENGTH)
    }

    this.name = this.name.trim()
  }
}

export class IdeaShard {
  /**
   * A constructor for idea shards that make up an idea set.
   * @param {DieItem} name - the name of the item shard/element of the set
   * @param {Die} die - the instance of a Die object where the shard came from
   */
  constructor (item, die) {
    if (!(die instanceof Die) && !(item instanceof DieItem)) return

    this.shardId = item.itemId
    this.shardName = item.name
    this.dieName = die.name
    this.dieId = die.id
  }
}

export const atomicShrimpSampleDiceSet = [
  new Die({ name: 'Power source', items: [ new DieItem({ name: 'Manual' }), new DieItem({ name: 'Electric' }), new DieItem({ name: 'Clockwork' }), new DieItem({ name: 'Solar' }), new DieItem({ name: 'Wind' }), new DieItem({ name: 'Water' }) ] }),
  new Die({ name: 'Size', items: [ new DieItem({ name: 'Giant' }), new DieItem({ name: 'Mini' }), new DieItem({ name: 'Pocket' }), new DieItem({ name: 'Portable' }), new DieItem({ name: 'Inhabitable' }), new DieItem({ name: 'Wearable' }) ] }),
  new Die({ name: 'Material', items: [ new DieItem({ name: 'Metal' }), new DieItem({ name: 'Wood' }), new DieItem({ name: 'Plastic' }), new DieItem({ name: 'Edible' }), new DieItem({ name: 'Organic' }), new DieItem({ name: 'Paper' }) ] }),
  new Die({ name: 'Device', items: [ new DieItem({ name: 'Robot' }), new DieItem({ name: 'Vehicle' }), new DieItem({ name: 'Game' }), new DieItem({ name: 'Tool' }), new DieItem({ name: 'Art' }), new DieItem({ name: 'Computer' }) ] }),
  new Die({ name: 'Scope', items: [ new DieItem({ name: 'Family' }), new DieItem({ name: 'Personal' }), new DieItem({ name: 'Office' }), new DieItem({ name: 'Industrial' }), new DieItem({ name: 'Public' }), new DieItem({ name: 'Home' }) ] }),
  new Die({ name: 'Implementation type', items: [ new DieItem({ name: 'Random' }), new DieItem({ name: 'Flying' }), new DieItem({ name: 'Underwater' }), new DieItem({ name: 'Stealth' }), new DieItem({ name: 'Disposable' }), new DieItem({ name: 'Self-build' }) ] })
]

export const diceStorage = {
  fetch: function () {
    const dice = []
    return diceStorageInstance.iterate(function (value, key, index) {
      dice.push(new Die(value))
    })
      .then(function () { return Promise.resolve(dice) })
      .catch(function (err) {
        console.error(err)
        return Promise.reject(err)
      })
  },
  save: function (dice) {
    diceStorageInstance.clear()

    for (const die of dice) {
      diceStorageInstance.setItem(String(die.id), die)
    }
  }
}

export const ideaStorage = {
  fetch: function () {
    const ideas = []
    return ideasStorageInstance.iterate(function (value, key, index) {
      ideas.push(new IdeaSet(value))
    })
      .then(function () {
        return Promise.resolve(ideas)
      })
      .catch(function (err) {
        return Promise.reject(err)
      })
  },
  save: function (ideas) {
    ideasStorageInstance.clear()

    for (const idea of ideas) {
      ideasStorageInstance.setItem(String(idea.id), idea)
    }
  }
}
